### PR TITLE
witness retention limits

### PR DESCRIPTION
**Pull request by evanmcc**
This PR establishes a chain var controlled limit on how many witnesses are allowed to be stored per gateway.  The limit is enforced by random selection of a subset of the witnesses.

### CODE REVIEW

File: src/transactions/v1/blockchain_txn_vars_v1.erl [hunk 1]
```diff
@@ -724,6 +724,8 @@ validate_var(?poc_path_limit, Value) ->
     validate_int(Value, "poc_path_limit", 3, 10, false);
 validate_var(?poc_witness_consideration_limit, Value) ->
     validate_int(Value, "poc_witness_consideration_limit", 10, 100, false);
+validate_var(?poc_witness_retention_limit, Value) ->
+    validate_int(Value, "poc_witness_consideration_limit", 10, 100, false);
```
Comment (vihu) [inline src/transactions/v1/blockchain_txn_vars_v1.erl:728]:
```suggestion
    validate_int(Value, "poc_witness_retention_limit", 10, 100, false);
```

File: src/ledger/v1/blockchain_ledger_gateway_v2.erl [hunk 2]
```diff
@@ -403,6 +403,11 @@ has_witness(#gateway_v2{witnesses=Witnesses}, WitnessAddr) ->
 witnesses(Gateway) ->
     Gateway#gateway_v2.witnesses.
 
+-spec witnesses(#{libp2p_crypto:pubkey_bin() => gateway_witness()}, gateway()) ->
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_gateway_v2.erl:406]: Uhm, this spec seems incorrect? The return type is a `gateway`

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 3]
```diff
@@ -1161,7 +1161,17 @@ add_gateway_witnesses(GatewayAddress, WitnessInfo, Ledger) ->
                                               erlang:error({add_gateway_error, Reason})
                                       end
                               end, GW0, WitnessInfo),
-            update_gateway(GW1, GatewayAddress, Ledger)
+            GW2 = case config(?poc_witness_retention_limit, Ledger) of
+                      {ok, Limit} ->
+                          Witnesses = blockchain_ledger_gateway_v2:witnesses(GW1),
+                          {ok, Height} = current_height(Ledger),
+                          %% this doesn't need to be hard to predict, just random
+                          RandState = blockchain_utils:rand_state(integer_to_binary(Height)),
```
Comment (Vagabond) [inline src/ledger/v1/blockchain_ledger_v1.erl:1169]: Would it be more efficent to check if the gateway is over the limit first?

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 4]
```diff
@@ -1161,7 +1161,17 @@ add_gateway_witnesses(GatewayAddress, WitnessInfo, Ledger) ->
...
+                          Witnesses1 = blockchain_utils:deterministic_subset(Limit, RandState, Witnesses),
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_v1.erl:1170]: `deteministic_subset` expects a list, `Witnesses` is a map

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 5]
```diff
@@ -1161,7 +1161,17 @@ add_gateway_witnesses(GatewayAddress, WitnessInfo, Ledger) ->
...
+                          blockchain_ledger_gateway_v2:witnesses(Witnesses1, GW1);
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_v1.erl:1171]: `witnesses/2` expects a map as first arg

**Comment by Vagabond** [comment]
What if, instead, we just reduced the rate of challenging/pathing through hotspots with more than X witnesses?

**Comment by evanmcc** [comment]
So, we had an offline conversation, and are a bit worried about losing this information.  I think that #504 will take care of this sufficiently for now, so I'm going to close it but not delete the branch for now, since we might need it in the future.

**Pull request closed without merging**